servicetalk-grpc-internal/src/main/java/io/servicetalk/grpc/internal/GrpcStatusUtils.java

```diff
@@ -0,0 +1,176 @@
+                    if (isEscapingChar(messageBytes[j])) {
+                        // Character that needs escaping found, continue on slow path with encoding.
+                        headers.set(GRPC_STATUS_MESSAGE, encodeMessage(messageBytes, j));
+                        return;
+                    }
+                }
+
+                // The UTF-8 encoded form doesn't require percent encoding. Use the ISO-8859-1 charset to preserve
+                // the byte representation of the string.
+                headers.set(GRPC_STATUS_MESSAGE, new String(messageBytes, StandardCharsets.ISO_8859_1));
+                return;
+            }
+        }
+        headers.set(GRPC_STATUS_MESSAGE, message);
+    }
+
+    /**
+     * Tries to read the status message from the {@link HttpHeaders} and percent-decode if necessary.
+     *
+     * @param headers the headers to load and decode the status message from.
+     * @return the decoded status message, or null of no message found in the header provided.
+     */
+    @Nullable
+    public static CharSequence getStatusMessage(final HttpHeaders headers) {
+        final CharSequence message = headers.get(GRPC_STATUS_MESSAGE);
+        if (message == null) {
+            return null;
+        }
+
+        final byte[] messageBytes = message.toString().getBytes(StandardCharsets.UTF_8);
+        for (int i = 0; i < messageBytes.length; i++) {
+            byte b = messageBytes[i];
+            if (b < ' ' || b >= '~' || (b == '%' && i + 2 < messageBytes.length)) {
+                return decodeMessage(messageBytes);
+            }
+        }
+        return message;
+    }
+
+    /**
+     * Decodes the {@link CharSequence} removing the percent-encoding where needed.
+     *
+     * @param messageBytes the message to decode.
+     * @return the deocded message.
+     */
+    private static CharSequence decodeMessage(final byte[] messageBytes) {
+        ByteBuffer buf = ByteBuffer.allocate(messageBytes.length);
+        for (int i = 0; i < messageBytes.length;) {
+            if (messageBytes[i] == '%' && i + 2 < messageBytes.length) {
+                try {
+                    buf.put((byte) Integer.parseInt(new String(messageBytes, i + 1, 2, StandardCharsets.US_ASCII), 16));
+                    i += 3;
+                    continue;
+                } catch (NumberFormatException e) {
+                    // ignore, fall through, just push the bytes.
+                }
+            }
+            buf.put(messageBytes[i]);
+            i += 1;
+        }
+        return new String(buf.array(), 0, buf.position(), StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Describes the character ranges which need escaping (essentially non-printable characters).
+     *
+     * @param b the character to check.
+     * @return true if it needs escaping, false otherwise.
+     */
+    private static boolean isEscapingChar(byte b) {
+        return b < ' ' || b >= '~' || b == '%';
+    }
```

Comment:
Fun fact: this may cause unnecessary encodings due to the signed nature of bytes because the byte 128 needs not be encoded but will be. The values 0x80-0xff are apparently legal for HTTP https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6
although it's not obvious if they're allowing it per the rules for gRPC. 🤷

Reply:
See comment above: the `b >= '~'` would have captured the values had they been unsigned so rolling them around to negative just means they'll get caught by `b < ' '` instead.